### PR TITLE
Invoke Epoch heartbeat in system chunk

### PR DESCRIPTION
**Pull request by jordanschalm**
This PR modifies the transaction script executed during the system chunk to invoke the `advanceBlock` method of the Epoch Heartbeat. This is necessary for epoch-related logic to execute every block.

### CODE REVIEW

**Review by arrivets — APPROVED** [review]

**Review by joshuahannan — APPROVED** [review]

File: fvm/serviceaccount.go [hunk 1]
```diff
@@ -7,15 +7,13 @@ import (
...
   prepare(serviceAccount: AuthAccount) { 
-
-  }
-
-  execute {
-    // TODO: replace with call to service account heartbeat
- 	log("pulse")
+	let heartbeat = serviceAccount.borrow<&FlowEpoch.Heartbeat>(from: FlowEpoch.heartbeatStoragePath)
+      ?? panic("Could not borrow heartbeat from storage path")
+    heartbeat.advanceBlock()
```
Comment (m4ksio) [inline fvm/serviceaccount.go:16]: Is there any way we can tests if this works correctly?

File: module/epochs/epoch_config.go [hunk 2]
```diff
@@ -30,10 +30,13 @@ type EpochConfig struct {
...
-		NumViewsInEpoch:              cadence.UInt64(100),
-		NumViewsInStakingAuction:     cadence.UInt64(10),
-		NumViewsInDKGPhase:           cadence.UInt64(10),
+		CurrentEpochCounter:      cadence.UInt64(0),
+		NumViewsInEpoch:          cadence.UInt64(10_000),
+		NumViewsInStakingAuction: cadence.UInt64(1000),
+		NumViewsInDKGPhase:       cadence.UInt64(1000),
+		//NumViewsInEpoch:              cadence.UInt64(100),
+		//NumViewsInStakingAuction:     cadence.UInt64(10),
+		//NumViewsInDKGPhase:           cadence.UInt64(10),
```
Comment (m4ksio) [inline module/epochs/epoch_config.go:39]: snip snip

**Review by m4ksio — APPROVED** [review]

**Comment by jordanschalm** [comment]
Hey @m4ksio @yhassanzadeh13, summarizing the problems demonstrated by the [VN Happy Path tests](https://github.com/onflow/flow-go/blob/ff92e3b5a8edfb80eb797e8896c8c2a0bdea8fcd/engine/verification/test/assignerfetcher_test.go#L44) failing here:

* The [first two test cases](https://github.com/onflow/flow-go/blob/ff92e3b5a8edfb80eb797e8896c8c2a0bdea8fcd/engine/verification/test/assignerfetcher_test.go#L52-L78) (1 block, 1 ER, 1 chunk) always pass
* Subsequent test cases (with more than 1 block, ER, chunk) fail, details 👇 

These test cases fail when the epoch smart contract is configured such that [`advanceBlock`](https://github.com/onflow/flow-core-contracts/blob/feature/epochs/contracts/epochs/FlowEpoch.cdc#L303) will trigger an epoch phase transition from `Staking` phase to `Setup` phase. This happens when the epoch is configured with a sufficiently small [`NumViewsInStakingAuction`](https://github.com/onflow/flow-go/blob/e92d1783b146b29d60ebc72f77a31604e753b0cc/module/epochs/epoch_config.go#L35). The codepath in the smart contract is [here](https://github.com/onflow/flow-core-contracts/blob/feature/epochs/contracts/epochs/FlowEpoch.cdc#L310). 

(1) When this `Staking->Setup` phase transition codepath is executed, the VN happy path tests fail while the `verifier` engine is verifying a chunk. The logs indicate that the computed final state commitment does not match the final state commitment provided by the ER or the chunk data pack:
```
{"level":"warn","role":"verification","engine":"verifier","time":"2021-05-14T16:20:37Z","message":"final state commitment doesn't match for chunk index 0, expected [f4de6fcf6ad3325abfc0af7218c7f9637fdb8608a16a8bb43eb17e2b40f84d35] but computed [67ac656a65bdd6c1844e1d6c0854d47c6f1efd090f9a91bef7d25e3a99202f63]"}
```
💡 This log is for chunk index 0, which should not be the system chunk because the test fixture includes collections in the block, and the system chunk is the final chunk 🤔 

(2) We can cause the `Staking->Setup` phase transition codepath to _not_ be executed in these tests by configuring the epoch with a large [`NumViewsInStakingAuction`](https://github.com/onflow/flow-go/blob/e92d1783b146b29d60ebc72f77a31604e753b0cc/module/epochs/epoch_config.go#L35). When it is configured this way, the tests pass and there are no error/warn logs indicating inconsistent state commitments.

File: module/epochs/epoch_config.go (hunk 2 above)
Comment (jordanschalm) [reply module/epochs/epoch_config.go:39]: 👍  Will do, still using both sets of values for debugging

File: fvm/serviceaccount.go (hunk 1 above)
Comment (jordanschalm) [reply fvm/serviceaccount.go:16]: We have tests for the smart contract business logic using the emulator. Do you mean a test that the system chunk calls the `advanceBlock` function at all?

**Comment by jordanschalm** [comment]
`TestNetworkSetupBasic` is failing, but this is a problem from the feature `feature/epochs` which @arrivets is investigating, rather than from this PR. Merging.

**Pull request merged**